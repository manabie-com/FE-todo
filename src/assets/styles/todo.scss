.ToDo__container {
	border: 1px solid rgba(0, 0, 0, 0.13);
	border-radius: 8px;
	width: 500px;
	margin-left: auto;
	margin-right: auto;
	margin-top: 4rem;
	padding: 24px;
	box-shadow: 2px 2px 1px rgba(0, 0, 0, 0.09), 3px 2px 3px rgba(0, 0, 0, 0.05);

	@media (max-width: 600px) {
		width: unset;
	}

	.Todo__creation {
		display: flex;

		.Todo__input {
			flex: 1 1;
		}
	}

	.ToDo__list {
		display: flex;
		flex-direction: column;
		margin-top: 1.5rem;
		max-height: calc(100vh - 355px);
		overflow: auto;

		.ToDo__item {
			display: flex;
			align-items: center;
			justify-content: space-between;

			> span {
				flex: 1 1;
				text-align: left;
				margin-left: 8px;
				cursor: pointer;
			}
			.ToDo__item_input {
				flex: 1 1;
				text-align: left;
				margin-left: 8px;
				margin: 0 15px;
			}
		}
	}

	.Todo__content {
		flex: 1 1;
	}

	.Todo__delete {
		outline: none;
		border: none;
		width: 32px;
		height: 32px;
		min-width: auto;
		min-height: auto;
		box-shadow: none;
		border-radius: 50%;
		display: flex;
		justify-content: center;
		align-items: center;
		cursor: pointer;
		color: rgb(220, 0, 78);
	}

	.Todo__action,
	.Todo__delete {
		width: 24px;
		height: 24px;
		flex-shrink: 0;
	}

	.Todo__toolbar {
		display: flex;
		align-items: center;
		justify-content: space-between;
		margin-top: 24px;
		flex-wrap: wrap;
		position: relative;

		> div:not(.Todo__tabs) {
			@media (max-width: 600px) {
				width: 100%;
				margin-top: 15px;
			}
		}
	}

	.Todo__tabs {
		display: flex;
		justify-content: center;
	}

	.Todo__tabs > *:not(:last-child) {
		margin-right: 8px;
	}

	.Action__btn {
		height: 32px;
		background-color: #e0e0e0;
	}
	.Active_btn {
		background-color: #1565c0;
		color: #fff;
	}
	.Clear_btn {
		background-color: #e53935;
		color: #fff;

		@media (max-width: 600px) {
			float: right;
		}
	}
}
