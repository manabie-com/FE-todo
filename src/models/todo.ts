// file name should be type, not model
export enum TodoStatus {
  ACTIVE = 'ACTIVE',
  COMPLETED = 'COMPLETED'
}

export interface Todo {
  id: string
  user_id: string
  content : string
  status?: TodoStatus
  created_date: string
  editable: boolean
}
