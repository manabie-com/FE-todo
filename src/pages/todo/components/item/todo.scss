@import "src/styles/colors.scss";

.todo-item {
  padding: 1rem;
  border-radius: 0.25rem;
  box-shadow: 0 0 5px 3px $color-gray;
  position: relative;
  margin-bottom: 0.5rem;

  .checkbox {
    width: 1.5rem;
    height: 1.5rem;
    margin-right: 0.5rem;
    top: 0.5rem;
    position: relative;
  }
  span {
    margin-right: 0.25rem;
  }
  button {
    position: absolute;
    right: 0.5rem;
    top: 0.5rem;
    padding: 0rem 0.4rem;
  }
}
