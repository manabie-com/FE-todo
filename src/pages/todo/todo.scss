.todo {
  width: 100%;
  max-width: 35rem;
  padding: 2rem 1rem;
  margin: auto;

  .form {
    width: 100%;
  }
}
