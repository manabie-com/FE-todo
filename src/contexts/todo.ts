import React from 'react';
import { AppState } from "../models/app";

interface CTodoContext {
  store: AppState,
  dispatch: Function
}


export const TodoContext = React.createContext({} as CTodoContext)
