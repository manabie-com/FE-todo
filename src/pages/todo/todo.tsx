import React, { useEffect, useReducer, useState } from "react";
import { RouteComponentProps } from "react-router-dom";
import "./todo.scss";
import reducer from "../../store/reducer";
import {
  setTodos,
  createTodo,
  deleteTodo,
  deleteAllTodos,
  updateTodoStatus,
  updateTodo,
  AppActions,
} from "../../store/actions";
import Service from "../../services";
import { EnhanceTodoStatus, TodoStatus } from "../../models/todo";
import { getStateFromLocalStorage, isTodoCompleted } from "../../utils";
import Input from "src/components/input";
import Button from "src/components/button";
import TodoItem from "./components/item";
import { History } from "history";
import Toolbar from "./components/toolbar";

export const fetchTodoData = (dispatch: React.Dispatch<AppActions>) => () => {
  (async () => {
    const resp = await Service.getTodos();
    dispatch(setTodos(resp || []));
  })();
};

export const onCreateTodo = (
  history: History,
  dispatch: React.Dispatch<AppActions>
) => async (evt: React.KeyboardEvent<HTMLFormElement>) => {
  evt.preventDefault();
  const form = evt.target as HTMLFormElement;

  const formData = new FormData(form);
  const content = formData.get("content") as string;
  if (!content) return;
  form.reset();

  try {
    const resp = await Service.createTodo(content);
    dispatch(createTodo(resp));
  } catch (err) {
    if (err.response.status === 401) {
      history.push("/");
    }
  }
};

export const onDeleteAllTodo = (dispatch: React.Dispatch<AppActions>) => () => {
  dispatch(deleteAllTodos());
};

export const onChangeFilter = (
  setShowing: React.Dispatch<EnhanceTodoStatus>
) => (status: EnhanceTodoStatus) => {
  setShowing(status);
};

export const onUpdateTodoStatus = (dispatch: React.Dispatch<AppActions>) => (
  id: string,
  checked: boolean
) => {
  dispatch(updateTodoStatus(id, checked));
};

export const onUpdateTodo = (dispatch: React.Dispatch<AppActions>) => (
  id: string,
  content: string
) => {
  dispatch(updateTodo(id, content));
};

export const onDeleteTodoStatus = (dispatch: React.Dispatch<AppActions>) => (
  id: string
) => {
  dispatch(deleteTodo(id));
};

export const ToDo: React.FC<RouteComponentProps> = ({ history }) => {
  const [{ todos }, dispatch] = useReducer(reducer, getStateFromLocalStorage());
  const [showing, setShowing] = useState<EnhanceTodoStatus>("ALL");

  useEffect(fetchTodoData(dispatch), []);

  const handleCreateTodo = onCreateTodo(history, dispatch);
  const handleDeleteAllTodo = onDeleteAllTodo(dispatch);
  const handleChangeFilter = onChangeFilter(setShowing);
  const handleUpdateTodoStatus = onUpdateTodoStatus(dispatch);
  const handleUpdateTodo = onUpdateTodo(dispatch);
  const handleDeleteTodoStatus = onDeleteTodoStatus(dispatch);

  const showTodos = todos.filter((todo) => {
    switch (showing) {
      case TodoStatus.ACTIVE:
        return todo.status === TodoStatus.ACTIVE;
      case TodoStatus.COMPLETED:
        return todo.status === TodoStatus.COMPLETED;
      default:
        return true;
    }
  });

  const activeTodos = todos.reduce(function (accum, todo) {
    return isTodoCompleted(todo) ? accum : accum + 1;
  }, 0);
  const completedTodos = todos.length - activeTodos;

  return (
    <div className="todo">
      <form className="form" onSubmit={handleCreateTodo}>
        <Input placeholder="What need to be done?" name="content" />
        <Input hidden type="submit" />
      </form>

      <Toolbar
        onChange={handleChangeFilter}
        totalTodos={todos.length}
        completedTodos={completedTodos}
        activeTodos={activeTodos}
        showing={showing}
      />

      <div className="mt-4 mb-2">
        {showTodos.map((todo, index) => (
          <TodoItem
            key={index}
            id={todo.id}
            content={todo.content}
            checked={todo.status === TodoStatus.COMPLETED}
            onUpdateStatus={handleUpdateTodoStatus}
            onUpdate={handleUpdateTodo}
            onDelete={handleDeleteTodoStatus}
          />
        ))}
      </div>

      <Button
        className="inline-block"
        variant="danger"
        onClick={handleDeleteAllTodo}
      >
        Clear all todos
      </Button>
    </div>
  );
};

export default ToDo;
