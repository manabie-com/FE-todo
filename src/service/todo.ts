import { TodoStatus } from 'models/todo';

const setTodos = todos => {
  localStorage.setItem('todos', JSON.stringify(todos));
};

const getTodos = () => {
  const todos = JSON.parse(localStorage.getItem('todos')) ?? [];
  if (!Array.isArray(todos) || todos.length === 0) {
    setTodos([]);
  }
  return todos;
};

const todoService = {
  list: () =>
    new Promise(resolve => {
      setTimeout(() => {
        const todos = getTodos();
        resolve(todos);
      }, 200);
    }),
  create: body =>
    new Promise(resolve => {
      setTimeout(() => {
        const todos = getTodos();
        const newTodos = [body, ...todos];
        setTodos(newTodos);
        resolve(newTodos);
      }, 200);
    }),

  delete: todoId =>
    new Promise(resolve => {
      setTimeout(() => {
        const todos = getTodos();
        const newTodos = todos.filter(({ id }) => id !== todoId);
        setTodos(newTodos);
        resolve(newTodos);
      }, 200);
    }),
  update: (todoId, data) =>
    new Promise(resolve => {
      setTimeout(() => {
        const todos = getTodos();
        const newTodos = todos.map(todo => (todo.id === todoId ? data : todo));
        setTodos(newTodos);
        resolve(newTodos);
      }, 200);
    }),
  deleteAll: () =>
    new Promise(resolve => {
      setTimeout(() => {
        setTodos([]);
        resolve([]);
      }, 200);
    }),
  completeAll: () => {
    new Promise(resolve => {
      setTimeout(() => {
        const todos = getTodos();
        const newTodos = todos.map(todo => ({
          ...todo,
          status: TodoStatus.COMPLETED
        }));
        setTodos(newTodos);
        resolve(newTodos);
      }, 200);
    });
  },
  activeAll: () => {
    new Promise(resolve => {
      setTimeout(() => {
        const todos = getTodos();
        const newTodos = todos.map(todo => ({
          ...todo,
          status: TodoStatus.ACTIVE
        }));
        setTodos(newTodos);
        resolve(newTodos);
      }, 200);
    });
  }
};

export default todoService;
