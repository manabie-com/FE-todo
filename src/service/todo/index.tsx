import { Todo, TodoStatus } from "models/todo"
import shortid from "shortid"
import { ITodo } from "./types"
import { ResponseDataType } from "../types"

class TodoService extends ITodo {
  createTodo = async (content: string): Promise<ResponseDataType<Todo>> => {
    return Promise.resolve({
      isSuccess: true,
      data: {
        content: content,
        created_date: new Date().toISOString(),
        status: TodoStatus.ACTIVE,
        id: shortid(),
        user_id: "firstUser",
      } as Todo,
    })
  }

  async getTodos(): Promise<Todo[]> {
    let todoJSON = localStorage.getItem("todos")
    return todoJSON ? JSON.parse(todoJSON) : []
  }
}

export default new TodoService()
