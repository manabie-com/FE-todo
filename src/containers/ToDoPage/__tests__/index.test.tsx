import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ToDoPage } from '..';
import { configureAppStore } from 'store/configureStore';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { TodoStatus } from 'models/todo';

const history = createMemoryHistory();
const renderTodoPage = store =>
  render(
    <Provider store={store}>
      <Router history={history}>
        <ToDoPage />
      </Router>
    </Provider>,
  );

describe('<ToDoPage />', () => {
  let component: ReturnType<typeof renderTodoPage>;
  let store: ReturnType<typeof configureAppStore>;

  beforeEach(() => {
    localStorage.setItem('token', 'test');
    store = configureAppStore();
    component = renderTodoPage(store);
  });

  it('should match snapshot', () => {
    expect(component.container).toMatchSnapshot();
  });

  it('should handle filter button click', () => {
    const button = component.getByText('Show active');
    fireEvent.click(button);
    expect(store.getState().todoPage.filter).toBe(TodoStatus.ACTIVE);
  });

  it('should redirect to login page', () => {
    localStorage.clear();
    component.unmount();
    component = renderTodoPage(store);
    const { container } = component;
    expect(container).toMatchSnapshot();
  });
});
