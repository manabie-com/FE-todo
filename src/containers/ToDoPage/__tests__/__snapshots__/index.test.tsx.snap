// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToDoPage /> should match snapshot 1`] = `
.c4 {
  outline: none;
  cursor: pointer;
  border: none;
  font-weight: 700;
  padding: 8px 16px;
  border-radius: 6px;
  background-color: #e0e1e2;
  color: rgba(0,0,0,0.6);
}

.c4:hover,
.c4:focus {
  background-color: #cacbcd;
  color: rgba(0,0,0,0.8);
}

.c4:active {
  background-color: #babbbc;
  color: rgba(0,0,0,0.9);
}

.c5 {
  outline: none;
  cursor: pointer;
  border: none;
  font-weight: 700;
  padding: 8px 16px;
  border-radius: 6px;
  background-color: #2185d0;
  color: #fff;
}

.c5:hover,
.c5:focus {
  background-color: #1678c2;
}

.c5:active {
  background-color: #1a69a4;
}

.c1 {
  padding: 8px 16px;
  border-radius: 6px;
  width: 100%;
  outline: none;
  border: 1px solid rgba(34,36,38,0.15);
}

.c1:focus {
  border-color: #85b7d9;
}

.c2 {
  padding: 24px 0;
}

.c0 {
  box-shadow: 0 1px 2px 0 rgba(34,36,38,0.15);
  border: 1px solid rgba(34,36,38,0.15);
  margin: 200px auto;
  max-width: 600px;
  padding: 16px;
  border-radius: 6px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

<div>
  <div
    class="c0"
  >
    <form>
      <input
        class="c1"
        placeholder="What need to be done?"
        value=""
      />
    </form>
    <div
      class="c2"
    >
      <div>
        Nothing to show
      </div>
    </div>
    <div
      class="c3"
    >
      <button
        class="c4"
      >
        Show all
      </button>
      <button
        class="c4"
      >
        Show active
      </button>
      <button
        class="c4"
      >
        Show completed
      </button>
      <button
        class="c5"
      >
        Complete all
      </button>
      <button
        class="c5"
      >
        Delete all
      </button>
    </div>
  </div>
</div>
`;

exports[`<ToDoPage /> should redirect to login page 1`] = `<div />`;
